docs/DevGuide/TensorExpressions.md

```diff
@@ -0,0 +1,411 @@
+
+\f{align}{
+  g_{tt} &= - \alpha^2 + \beta^m \beta^n \gamma_{mn} \\
+  g_{ti} &= \gamma_{mi} \beta^m  \\
+  g_{ij} &= \gamma_{ij}
+\f}
+```
+// spatial_metric is type tnsr::ii<DataVector, 3>, shift is type
+// tnsr::I<DataVector, 3>, and lapse is type Scalar<DataVector>
+
+tnsr::aa<DataVector, 3> spacetime_metric{};
+tenex::evaluate<ti::t, ti::t>(
+    make_not_null(&spacetime_metric),
+    -lapse() * lapse() + shift(ti::M) * shift(ti::N) *
+                spatial_metric(ti::m, ti::n));
+tenex::evaluate<ti::t, ti::i>(
+    make_not_null(&spacetime_metric),
+    spatial_metric(ti::m, ti::i) * shift(ti::M));
+tenex::evaluate<ti::i, ti::j>(
+    make_not_null(&spacetime_metric), spatial_metric(ti::i, ti::j));
+```
+
+### Using the LHS Tensor in the RHS expression {#te_using_lhs_tensor_in_rhs}
+You can use the LHS `Tensor` in the RHS expression to emulate operations like
+`+=`, `*=`, etc. For example, say you would like to emulate the following:
+```
+// pseudocode
+L_ab = R_ab
+L_ab += + 2.0 * S_ba
+```
+You can do the operation in the 2nd line above by calling
+\ref tenex::update "update" instead of \ref tenex::evaluate "evaluate":
+```
+auto L = tenex::evaluate<ti::a, ti::b>(R(ti::a, ti::b));
+// use the LHS tensor in the RHS
+tenex::update<ti::a, ti::b>(
+    make_not_null(&L), L(ti::a, ti::b) + 2.0 * S(ti::b, ti::a));
+```
+One limitation is that when using the LHS tensor in the RHS expression, the
+index order used for the LHS tensor must be the same in the LHS and RHS, e.g.
+the following is not allowed and will yield a runtime error:
+```
+// ERROR: index order for L on LHS and RHS is not the same
+tenex::update<ti::a, ti::b>(
+    make_not_null(&L), L(ti::b, ti::a) + 2.0 * S(ti::b, ti::a));
+```
+
+**Note:** It is not advised to use very large RHS expressions with
+\ref tenex::update "update" because runtime performance does not scale well as
+the number of operations gets very large. This is because
+\ref tenex::evaluate "evaluate" breaks up large expressions into smaller ones,
+but \ref tenex::update "update" cannot. One way around this is to break up the
+expression and use more than one call to \ref tenex::update "update".
+
+# Compile time math checks {#te_compile_time_math_checks}
+For all operations, mathematical legality is checked at compile time. The
+compiler will catch what is not sound to write on paper, which includes things
+like no repeated indices, can't divide by a tensor with rank > 0, and that
+spatial dimensions, frames, valences, index types (spatial or spacetime), and
+ranks of tensors match where they should.
+
+Here are some examples of illegal math that the compiler will catch:
+
+```
+tnsr::ab<double, 3, Frame::Inertial> R{};
+tnsr::ab<double, 3, Frame::Inertial> S{};
+tnsr::ab<double, 3, Frame::Grid> T{};
+tnsr::AB<double, 2, Frame::Inertial> G{};
+
+// ERROR: LHS and RHS indices don't match
+auto result1 = tenex::evaluate<ti::a, ti::c>(R(ti::a, ti::b) + S(ti:a::ti::b));
```

Comment:
Could you add a blank line between the result line and the next "ERROR" comment?